cylc/rose/jinja2_parser.py

```diff
@@ -27,6 +29,155 @@
+        nonlocal instances  # record of uses of deprecated syntax
+        for lineno, token, value_str in stream:
+            if (
+                token == jinja2.lexer.TOKEN_INTEGER
+                and len(value_str) > 1
+                and value_str[0] == '0'
+            ):
+                instances.add(value_str)
+            yield (lineno, token, _strip_leading_zeros(value_str))
+
+    def _inner(
+        self,
+        stream,  # : t.Iterable[t.Tuple[int, str, str]],
+        name,  # : t.Optional[str] = None,
+        filename,  # : t.Optional[str] = None,
+    ):  # -> t.Iterator[Token]:
+        nonlocal fcn
+        return fcn(self, _stream(stream), name, filename)
+
+    _inner.__wrapped__ = fcn  # save the un-patched function
+    _inner._instances = instances  # save the set of uses of deprecated syntax
+
+    return _inner
+
+
+@contextmanager
+def patch_jinja2_leading_zeros():
+    """Back support integers with leading zeros in Jinja2 v3.
+
+    Jinja2 v3 dropped support for integers with leading zeros, these are
+    heavily used throughout Rose configurations. Since there was no deprecation
+    warning in Jinja2 v2 we have implemented this patch to extend support for
+    a short period to help our users to transition.
+
+    This patch will issue a warning if usage of the deprecated syntax is
+    detected during the course of its usage.
+
+    Warning:
+        This is a *global* patch applied to the Jinja2 library whilst the
+        context manager is open. Do not use this with parallel/async code.
+
+    Examples:
+        >>> env = NativeEnvironment()
+
+        The integer "1" is ok:
+        >>> env.parse('{{ 1 }}')
+        Template(body=[Output(nodes=[Const(value=1)])])
+
+        However "01" is no longer supported:
+        >>> env.parse('{{ 01 }}')
+        Traceback (most recent call last):
+        jinja2.exceptions.TemplateSyntaxError: expected token ...
+
+        The patch returns support (the leading-zero gets stripped):
+        >>> with patch_jinja2_leading_zeros():
+        ...     env.parse('{{ 01 }}')
+        Template(body=[Output(nodes=[Const(value=1)])])
+
+        The patch can handle any number of arbitrary leading zeros:
+        >>> with patch_jinja2_leading_zeros():
+        ...     env.parse('{{ 0000000001 }}')
+        Template(body=[Output(nodes=[Const(value=1)])])
+
+        Once the "with" closes we go back to vanilla Jinja2 behaviour:
+        >>> env.parse('{{ 01 }}')
+        Traceback (most recent call last):
+        jinja2.exceptions.TemplateSyntaxError: expected token ...
+
+    """
+    # clear any previously cashed lexer instances
+    jinja2.lexer._lexer_cache.clear()
+
+    # apply the code patch (new lexer instances will pick up these changes)
+    _integer_re = deepcopy(jinja2.lexer.integer_re)
+    jinja2.lexer.integer_re = re.compile(
+        rf'''
+            (
+                [0-9](_?\d)* # decimal
```

Comment:
This regex is a little confusing, may be worth commenting

```suggestion
                [0-9](_?\d)* # modified jinja2.lexer.integer_re to support leading 0s
```